### PR TITLE
Refactor everything relating to OCI client

### PR DESCRIPTION
**Pull request by zregvart**
This refactors all codepaths to use the single OCI client abstraction and a single fake/mock in testing.

Now all options provided to go-containerregistry are maintaned in the single place `internal/utils/oci/client.go`.

### CODE REVIEW

**Review by simonbaird — APPROVED** [review]
I'm reviewing the intention more than the code changes, but lgtm afaict. The tests are green so should be good right?

File: internal/rego/oci/oci.go [hunk 1]
```diff
@@ -226,7 +226,7 @@ func ociImageManifest(bctx rego.BuiltinContext, a *ast.Term) (*ast.Term, error)
 		remote.WithAuthFromKeychain(authn.DefaultKeychain),
 	}
 
-	image, err := oci.NewClient(bctx.Context).Image(ref, opts...)
+	image, err := oci.NewClient(bctx.Context, opts...).Image(ref)
```
Comment (lcarva) [inline internal/rego/oci/oci.go:229]: Maybe just use the default opts here as well?

File: internal/utils/testing_image.go [hunk 2]
```diff
@@ -14,7 +14,7 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-//go:build unit
+//go:build unit || integration
```
Comment (lcarva) [inline internal/utils/testing_image.go:17]: Oh interesting! Didn't realize this was a thing.

File: internal/rego/sigstore/sigstore_test.go [hunk 3]
```diff
@@ -63,7 +62,7 @@ func TestSigstoreVerifyImage(t *testing.T) {
 			uri:     ast.StringTerm(goodImage.String()),
 			opts:    options{ignoreRekor: true, publicKey: utils.TestPublicKey},
 			optsVerifier: func(args mock.Arguments) {
-				checkOpts := args.Get(2).(*cosign.CheckOpts)
+				checkOpts := args.Get(1).(*cosign.CheckOpts)
```
[the remark this answers is not in the document]
Comment (zregvart) [reply internal/rego/sigstore/sigstore_test.go:65]: Method signature is different, the Context is not passed as the first argument

File: internal/utils/oci/client.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+		return nil
+	}
+
+	if userCache, err := os.UserCacheDir(); err != nil {
+		log.Debug("unable to find user cache directory")
+		return nil
+	} else {
+		imgCacheDir := path.Join(userCache, "ec", "images")
+		if err := os.MkdirAll(imgCacheDir, 0700); err != nil {
+			log.Debugf("unable to create temporary directory for image cache in %q: %v", imgCacheDir, err)
```
[the remark this answers is not in the document]
Comment (zregvart) [reply internal/utils/oci/client.go:62]: Good catch!

File: internal/utils/oci/client.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+func createRemoteOptions(ctx context.Context) []remote.Option {
+	backoff := remote.Backoff{
+		Duration: 1.0 * time.Second,
+		Factor:   2.0,
+		Jitter:   0.1,
+		Steps:    3,
+	}
+
+	return []remote.Option{
+		imageRefTransport,
```
[the remark this answers is not in the document]
Comment (zregvart) [reply internal/utils/oci/client.go:78]: `imageRefTransport` is modified in tests (this code was moved there)

File: internal/rego/oci/oci.go (hunk 1 above)
Comment (zregvart) [reply internal/rego/oci/oci.go:229]: Yeah

File: internal/rego/oci/oci.go [hunk 6]
```diff
@@ -167,7 +167,7 @@ func ociBlob(bctx rego.BuiltinContext, a *ast.Term) (*ast.Term, error) {
 		remote.WithAuthFromKeychain(authn.DefaultKeychain),
 	}
 
-	rawLayer, err := oci.NewClient(bctx.Context).Layer(ref, opts...)
+	rawLayer, err := oci.NewClient(bctx.Context, opts...).Layer(ref)
```
Comment (lcarva) [inline internal/rego/oci/oci.go:170]: Should this one also use the default options?

**Review by lcarva — APPROVED** [review]

File: internal/applicationsnapshot/input.go [hunk 7]
```diff
@@ -248,7 +197,7 @@ func expandImageIndex(ctx context.Context, snap *app.SnapshotSpec) {
 			continue
 		}
 
-		desc, err := client.Get(ref)
+		desc, err := client.Head(ref)
```
[the remark this answers is not in the document]
Comment (lcarva) [reply internal/applicationsnapshot/input.go:200]: It looks like the media type is indeed in a header, and the library will [throw an error](https://github.com/google/go-containerregistry/blob/ff385a972813c79bbd5fc89357ff2cefe3e5b43c/pkg/v1/remote/fetcher.go#L215) if the header is missing. That's good enough for me!

**Comment by lcarva** [comment]
Thanks for doing this much needed re-factoring. It's beautiful :pinched_fingers:

**Pull request merged**